src/probnum/diffeq/steprule.py

```diff
@@ -19,66 +24,109 @@ def is_accepted(self, proposedstep, errorest, **kwargs):
-    localconvrate : float
-        (Estimated) convergence rate of the solver.
+    firststep : float
+        First step to be taken by the ODE solver (which happens in absence of error estimates).
     limitchange : list with 2 elements, optional
         Lower and upper bounds for computed change of step.
     safetyscale : float, optional
         Safety factor for proposal of distributions, 0 << safetyscale < 1
+    minstep : float, optional
+        Minimum step that is allowed. A runtime error is thrown if the proposed step is smaller. Default is 1e-15.
+    maxstep : float, optional
+        Maximum step that is allowed. A runtime error is thrown if the proposed step is larger. Default is 1e15.
     """
 
     def __init__(
         self,
-        tol_per_step,
-        localconvrate,
+        firststep,
         limitchange=(0.1, 5.0),
         safetyscale=0.95,
-        **kwargs
+        minstep=1e-15,
+        maxstep=1e15,
     ):
-        self.tol_per_step = float(tol_per_step)
         self.safetyscale = float(safetyscale)
-        self.localconvrate = float(localconvrate + 1)
         self.limitchange = limitchange
+        self.minstep = minstep
+        self.maxstep = maxstep
+
+        super().__init__(firststep=firststep)
 
-    def suggest(self, laststep, errorest, **kwargs):
+    def suggest(self, laststep, scaled_error, localconvrate=None):
         small, large = self.limitchange
-        ratio = self.tol_per_step / (laststep * errorest)
-        change = self.safetyscale * ratio ** (1.0 / self.localconvrate)
+
+        ratio = 1.0 / scaled_error
+        change = self.safetyscale * ratio ** (1.0 / localconvrate)
+
+        # The below code should be doable in a single line?
         if change < small:
             step = small * laststep
         elif large < change:
             step = large * laststep
         else:
             step = change * laststep
-        if step < 1e-15:
-            print("Warning: Stepsize is num. zero (h=%.1e)" % step)
+
+        if step < self.minstep:
+            raise RuntimeError("Step-size smaller than minimum step-size")
+        if step > self.maxstep:
+            raise RuntimeError("Step-size larger than maximum step-size")
         return step
 
-    def is_accepted(self, proposedstep, errorest, **kwargs):
-        return errorest * proposedstep < self.tol_per_step
+    # Looks unnecessary, though maybe we want tolerance per unit step
+    # in which case it is good to have such method in here.
+    def is_accepted(self, laststep, scaled_error, localconvrate=None):
+        return scaled_error < 1
+
+    # In here, because (i) we do not want to compute it for constant steps,
+    # and in fact, we don't even want to think about which value atol and rtol should have;
+    # (ii) having it in StepRule makes it easier to test, because the class is more light-weight;
+    # (iii) who knows, maybe there are other ways of dealing with this.
+    def errorest_to_internalnorm(
```

Comment:
What do you think about moving this to the base class? This scaling would also be needed in a PI-controller.

Reply:
Ok I think I understand why: You overwrite it in the constant step rule. I guess the additional computational effort is quite low so that it should not matter too much if it gets called during constant steps, so I think I'd still be in favor of having it in the base class. Optimizing the constant steps requires a bit more work anyways, for example the local errors are computed in the solver step but are never used. What do you think?

Reply:
Not sure. Another reason for doing it like this would be that there are ode solvers which cannot compute error estimates (hello particle filter), and as such, are only usable with constant steps. In these cases I tend to return `errorest=np.nan` which would lead to an error if this one here is computed for constant steps. This, and the reasons I explained in the comments below make me think that it is rather well-suited here. I would be lying though, if I said that I think it is clearly better...